File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.cc
#include <iostream>
#include <queue>

using namespace std;

int arr[1001][1001];
int visited[1001][1001]={0};
int x,y;
int n,m;
int dx[4]={1,-1,0,0};
int dy[4]={0,0,1,-1};

void bfs(int start, int end){
  queue<pair<int,int>> q;
  q.push({start, end});
  visited[start][end]=1;
  arr[start][end]=0;
  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      if(arr[i][j]==1){
        while(!q.empty()){
        int front=q.front().first;
        int back=q.front().second;
        q.pop();

        for(int k=0;k<4;k++){
          int nx=front+dx[k];
          int ny=back+dy[k];

          if(arr[nx][ny]==1 && visited[nx][ny]==0 && nx>=0 && ny>=0 && nx<n && ny<m){
            q.push({nx,ny});
            arr[nx][ny]=arr[front][back]+1;
            visited[nx][ny]=1;
          }
        }
      }
      }
      
    }
  }
}
int main(){
  cin>>n>>m;
  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      cin>>arr[i][j];
      if(arr[i][j]==2){
        x=i;
        y=j;
      }
    }
  }
  bfs(x,y);
  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      if(visited[i][j]==0 && arr[i][j]!=0){
        arr[i][j]=-1;
      }
    }
  }
  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      cout<<arr[i][j]<<" ";
    }
    cout<<endl;
  }

  
}

File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.java

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.*;
public class Main {
    static int[] dx = {0, 0,1,-1};
    static int[] dy = {1 , -1,0,0};
    static int[][] arr;
    static boolean[][] visited;
    static int[][] map;
    static int n;
    static int m;
    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        String[] docs = br.readLine().split(" ");
        n = Integer.parseInt(docs[0]);
        m = Integer.parseInt(docs[1]);
        arr = new int[n][m];
        visited = new boolean[n][m];
        map = new int[n][m];
        int x = 0;
        int y = 0;
        for(int i = 0; i < n; i++){
            String[] tmp = br.readLine().split(" ");
            for(int j = 0; j < m; j++){
                arr[i][j] = Integer.parseInt(tmp[j]);
                if(arr[i][j] == 2){
                    x = i;
                    y = j;
                }
            }
        }
        bfs(x, y);
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(arr[i][j] == 1 && !visited[i][j]){
                    map[i][j] = -   1;
                }
            }
        }
        bw.flush();
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                bw.write(map[i][j] + " ");
            }
            bw.write("\n");
        }
        bw.close();
    }

    public static void bfs(int x, int y){
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(x, y));
        visited[x][y] = true;
        while(!q.isEmpty()){
            Pair p = q.poll();
            int thisx = p.x;
            int thisy = p.y;
            for(int i=0;i<4;i++){
                int nx = thisx + dx[i];
                int ny = thisy + dy[i];
                if(nx >= 0 && nx < n && ny >= 0 && ny < m && !visited[nx][ny] && arr[nx][ny] == 1){
                    map[nx][ny] = map[thisx][thisy] + 1;
                    visited[nx][ny] = true;
                    q.add(new Pair(nx, ny));
                }
            }

        }
    }

    public static class Pair{
        int x;
        int y;
        Pair(int x, int y){
            this.x = x;
            this.y = y;
        }
    }
}
